c++ chrono operator""ms error

# Stack trace

```
08:00:35,168 INFO  MeasurementHolder CXXLanguageFrontend: Parsing sourcefile main.cpp
08:00:35,567 INFO  CXXLanguageFrontend Parsed 281 bytes in main.cpp corresponding roughly to 5 LoC
08:00:35,580 INFO  MeasurementHolder CXXLanguageFrontend: Parsing sourcefile main.cpp done in 399 ms
08:00:35,580 INFO  MeasurementHolder CXXLanguageFrontend: Transforming main.cpp to CPG
08:00:35,621 INFO  MeasurementHolder TranslationManager: Translation into full graph done in 646 ms
java.util.concurrent.ExecutionException: java.lang.NumberFormatException: For input string: "250ms"
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
	at de.fraunhofer.aisec.cpg_vis_neo4j.Application.call(Application.kt:366)
	at de.fraunhofer.aisec.cpg_vis_neo4j.Application.call(Application.kt:66)
	at picocli.CommandLine.executeUserObject(CommandLine.java:2041)
	at picocli.CommandLine.access$1500(CommandLine.java:148)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2461)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2453)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2415)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2273)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2417)
	at picocli.CommandLine.execute(CommandLine.java:2170)
	at de.fraunhofer.aisec.cpg_vis_neo4j.ApplicationKt.main(Application.kt:406)
Caused by: java.lang.NumberFormatException: For input string: "250ms"
	at java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:67)
	at java.base/java.lang.Integer.parseInt(Integer.java:668)
	at java.base/java.math.BigInteger.<init>(BigInteger.java:538)
	at de.fraunhofer.aisec.cpg.frontends.cpp.ExpressionHandler.handleIntegerLiteral(ExpressionHandler.kt:657)
	at de.fraunhofer.aisec.cpg.frontends.cpp.ExpressionHandler.handleLiteralExpression(ExpressionHandler.kt:490)
	at de.fraunhofer.aisec.cpg.frontends.cpp.ExpressionHandler.handleNode(ExpressionHandler.kt:58)
	at de.fraunhofer.aisec.cpg.frontends.cpp.ExpressionHandler.handleNode(ExpressionHandler.kt:53)
	at de.fraunhofer.aisec.cpg.frontends.cpp.CXXHandler.handle(CXXHandler.kt:56)
	at de.fraunhofer.aisec.cpg.frontends.cpp.InitializerHandler.handleEqualsInitializer(InitializerHandler.kt:68)
	at de.fraunhofer.aisec.cpg.frontends.cpp.InitializerHandler.handleNode(InitializerHandler.kt:43)
	at de.fraunhofer.aisec.cpg.frontends.cpp.InitializerHandler.handleNode(InitializerHandler.kt:38)
	at de.fraunhofer.aisec.cpg.frontends.cpp.CXXHandler.handle(CXXHandler.kt:56)
	at de.fraunhofer.aisec.cpg.frontends.cpp.DeclaratorHandler.handleDeclarator(DeclaratorHandler.kt:109)
	at de.fraunhofer.aisec.cpg.frontends.cpp.DeclaratorHandler.handleNode(DeclaratorHandler.kt:58)
	at de.fraunhofer.aisec.cpg.frontends.cpp.DeclaratorHandler.handleNode(DeclaratorHandler.kt:51)
	at de.fraunhofer.aisec.cpg.frontends.cpp.CXXHandler.handle(CXXHandler.kt:56)
	at de.fraunhofer.aisec.cpg.frontends.cpp.DeclarationHandler.handleSimpleDeclaration(DeclarationHandler.kt:501)
	at de.fraunhofer.aisec.cpg.frontends.cpp.DeclarationHandler.handleNode(DeclarationHandler.kt:63)
	at de.fraunhofer.aisec.cpg.frontends.cpp.DeclarationHandler.handleNode(DeclarationHandler.kt:58)
	at de.fraunhofer.aisec.cpg.frontends.cpp.CXXHandler.handle(CXXHandler.kt:56)
	at de.fraunhofer.aisec.cpg.frontends.cpp.StatementHandler.handleDeclarationStatement(StatementHandler.kt:292)
	at de.fraunhofer.aisec.cpg.frontends.cpp.StatementHandler.handleNode(StatementHandler.kt:52)
```


# Unexpected behavior

Since #1061 parsing a chrono `ms` values result in a crash.
The cpg seems to convert the ms literal to an integer, which fails.
Pre #1061, the example code could be parsed without error.
The value of the `ms` literal was `<error>`.


# Code of interest

```c++
#include <iostream>
#include <chrono>

int main()
{
    using namespace std::chrono_literals;
    auto d1 = 250ms;
    std::chrono::milliseconds d2 = 1s;
    std::cout << "250ms = " << d1.count() << " milliseconds\n"
              << "1s = " << d2.count() << " milliseconds\n";
    return 0;
}
```


# Extensions to the library, e.g. added or changed LanguageFrontends or Passes

None


# Changes to the configuration

None

--- a/cpg-core/src/main/java/de/fraunhofer/aisec/cpg/graph/types/BooleanType.kt
+++ b/cpg-core/src/main/java/de/fraunhofer/aisec/cpg/graph/types/BooleanType.kt
@@ -1,0 +1,42 @@
+/*
+ * Copyright (c) 2023, Fraunhofer AISEC. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ *                    $$$$$$\  $$$$$$$\   $$$$$$\
+ *                   $$  __$$\ $$  __$$\ $$  __$$\
+ *                   $$ /  \__|$$ |  $$ |$$ /  \__|
+ *                   $$ |      $$$$$$$  |$$ |$$$$\
+ *                   $$ |      $$  ____/ $$ |\_$$ |
+ *                   $$ |  $$\ $$ |      $$ |  $$ |
+ *                   \$$$$$   |$$ |      \$$$$$   |
+ *                    \______/ \__|       \______/
+ *
+ */
+package de.fraunhofer.aisec.cpg.graph.types
+
+import de.fraunhofer.aisec.cpg.frontends.Language
+import de.fraunhofer.aisec.cpg.frontends.LanguageFrontend
+
+/** Instances of this class represent boolean types. */
+class BooleanType(
+    typeName: CharSequence = "bool",
+    bitWidth: Int? = 1,
+    language: Language<out LanguageFrontend>? = null,
+    modifier: Modifier = Modifier.UNSIGNED
+) : NumericType(typeName, bitWidth, language, modifier) {
+
+    override fun duplicate(): Type {
+        return BooleanType(this.name, bitWidth, language, modifier)
+    }
+}

--- a/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/cpp/CPPLanguage.kt
+++ b/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/cpp/CPPLanguage.kt
@@ -56,7 +56,7 @@ class CPPLanguage :
     @Transient
     override val simpleTypes =
         mapOf(
-            "boolean" to IntegerType("boolean", 1, this, NumericType.Modifier.SIGNED),
+            "bool" to BooleanType("bool", 1, this, NumericType.Modifier.SIGNED),
             "char" to IntegerType("char", 8, this, NumericType.Modifier.NOT_APPLICABLE),
             "byte" to IntegerType("byte", 8, this, NumericType.Modifier.SIGNED),
             "short" to IntegerType("short", 16, this, NumericType.Modifier.SIGNED),
@@ -79,6 +79,8 @@ class CPPLanguage :
             "unsigned int" to IntegerType("unsigned int", 32, this, NumericType.Modifier.UNSIGNED),
             "unsigned long" to
                 IntegerType("unsigned long", 64, this, NumericType.Modifier.UNSIGNED),
+            "unsigned long long" to
+                IntegerType("unsigned long long", 64, this, NumericType.Modifier.UNSIGNED),
             "unsigned long long int" to
                 IntegerType("unsigned long long int", 64, this, NumericType.Modifier.UNSIGNED),
             "std::string" to StringType("std::string", this),

--- a/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/cpp/ExpressionHandler.kt
+++ b/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/cpp/ExpressionHandler.kt
@@ -627,24 +627,20 @@ class ExpressionHandler(lang: CXXLanguageFrontend) :
         return die
     }
 
-    private fun handleIntegerLiteral(ctx: IASTLiteralExpression): Literal<*> {
-        val value = String(ctx.value).lowercase(Locale.getDefault())
+    private fun handleIntegerLiteral(ctx: IASTLiteralExpression): Expression {
+        var (strippedValue, suffix) = ctx.valueWithSuffix
         val bigValue: BigInteger
-        val suffix = value.suffix
-
-        // first, strip the suffix from the value
-        var strippedValue = value.substring(0, value.length - suffix.length)
 
         // next, check for possible prefixes
         var radix = 10
         var offset = 0
-        if (value.startsWith("0b")) {
+        if (strippedValue.startsWith("0b")) {
             radix = 2 // binary
             offset = 2 // len("0b")
-        } else if (value.startsWith("0x")) {
+        } else if (strippedValue.startsWith("0x")) {
             radix = 16 // hex
             offset = 2 // len("0x")
-        } else if (value.startsWith("0") && strippedValue.length > 1) {
+        } else if (strippedValue.startsWith("0") && strippedValue.length > 1) {
             radix = 8 // octal
             offset = 1 // len("0")
         }
@@ -656,15 +652,32 @@ class ExpressionHandler(lang: CXXLanguageFrontend) :
         strippedValue = strippedValue.replace("'", "")
 
         // basically we parse everything as BigInteger and then decide what to do
-        bigValue = BigInteger(strippedValue, radix)
-        val numberValue: Number
-        if ("ull" == suffix || "ul" == suffix) {
-            // unsigned long (long) will always be represented as BigInteger
-            numberValue = bigValue
-        } else if ("ll" == suffix || "l" == suffix) {
-            // both long and long long can be represented in Java long, but only within
-            // Long.MAX_VALUE
-            if (bigValue > BigInteger.valueOf(Long.MAX_VALUE)) {
+        try {
+            bigValue = BigInteger(strippedValue, radix)
+            val numberValue: Number
+            if ("ull" == suffix || "ul" == suffix) {
+                // unsigned long (long) will always be represented as BigInteger
+                numberValue = bigValue
+            } else if ("ll" == suffix || "l" == suffix) {
+                // both long and long long can be represented in Java long, but only within
+                // Long.MAX_VALUE
+                if (bigValue > BigInteger.valueOf(Long.MAX_VALUE)) {
+                    // keep it as BigInteger
+                    numberValue = bigValue
+                    Util.warnWithFileLocation(
+                        frontend,
+                        ctx,
+                        log,
+                        "Integer literal {} is too large to be represented in a signed type, interpreting it as unsigned.",
+                        ctx
+                    )
+                } else {
+                    numberValue = bigValue.toLong()
+                }
+            } else if (bigValue > BigInteger.valueOf(Long.MAX_VALUE)) {
+                // No suffix, we just cast it to the appropriate signed type that is required, but
+                // only within Long.MAX_VALUE
+
                 // keep it as BigInteger
                 numberValue = bigValue
                 Util.warnWithFileLocation(
@@ -674,59 +687,44 @@ class ExpressionHandler(lang: CXXLanguageFrontend) :
                     "Integer literal {} is too large to be represented in a signed type, interpreting it as unsigned.",
                     ctx
                 )
-            } else {
+            } else if (bigValue.toLong() > Int.MAX_VALUE) {
                 numberValue = bigValue.toLong()
-            }
-        } else if (bigValue > BigInteger.valueOf(Long.MAX_VALUE)) {
-            // No suffix, we just cast it to the appropriate signed type that is required, but only
-            // within Long.MAX_VALUE
-
-            // keep it as BigInteger
-            numberValue = bigValue
-            Util.warnWithFileLocation(
-                frontend,
-                ctx,
-                log,
-                "Integer literal {} is too large to be represented in a signed type, interpreting it as unsigned.",
-                ctx
-            )
-        } else if (bigValue.toLong() > Int.MAX_VALUE) {
-            numberValue = bigValue.toLong()
-        } else {
-            numberValue = bigValue.toInt()
-        }
-
-        // retrieve type based on stored Java number
-        val type =
-            if (numberValue is BigInteger && "ul" == suffix) {
-                // we follow the way clang/llvm handles this and this seems to always
-                // be an unsigned long long, except if it is explicitly specified as ul
-                parseType("unsigned long")
-            } else if (numberValue is BigInteger) {
-                parseType("unsigned long long")
-            } else if (numberValue is Long && "ll" == suffix) {
-                // differentiate between long and long long
-                parseType("long long")
-            } else if (numberValue is Long) {
-                parseType("long")
             } else {
-                parseType("int")
+                numberValue = bigValue.toInt()
             }
 
-        return newLiteral(numberValue, type, ctx.rawSignature)
+            // retrieve type based on stored Java number
+            val type =
+                if (numberValue is BigInteger && "ul" == suffix) {
+                    // we follow the way clang/llvm handles this and this seems to always
+                    // be an unsigned long long, except if it is explicitly specified as ul
+                    parseType("unsigned long")
+                } else if (numberValue is BigInteger) {
+                    parseType("unsigned long long")
+                } else if (numberValue is Long && "ll" == suffix) {
+                    // differentiate between long and long long
+                    parseType("long long")
+                } else if (numberValue is Long) {
+                    parseType("long")
+                } else {
+                    parseType("int")
+                }
+
+            return newLiteral(numberValue, type, ctx.rawSignature)
+        } catch (ex: NumberFormatException) {
+            // It could be that we cannot parse the literal, in this case we return an error
+            return ProblemExpression("could not parse literal: ${ex.message}")
+        }
     }
 
     private fun handleFloatLiteral(ctx: IASTLiteralExpression): Expression {
-        val value = String(ctx.value).lowercase(Locale.getDefault())
-        val suffix = value.suffix
+        val (strippedValue, suffix) = ctx.valueWithSuffix
 
-        // first, strip the suffix from the value
-        val strippedValue = value.substring(0, value.length - suffix.length)
-
-        return if (suffix == "f") {
-            newLiteral(strippedValue.toFloat(), parseType("float"), ctx.rawSignature)
-        } else {
-            newLiteral(strippedValue.toDouble(), parseType("double"), ctx.rawSignature)
+        return when (suffix) {
+            "f" -> newLiteral(strippedValue.toFloat(), parseType("float"), ctx.rawSignature)
+            "l" ->
+                newLiteral(strippedValue.toBigDecimal(), parseType("long double"), ctx.rawSignature)
+            else -> newLiteral(strippedValue.toDouble(), parseType("double"), ctx.rawSignature)
         }
     }
 
@@ -746,20 +744,25 @@ class ExpressionHandler(lang: CXXLanguageFrontend) :
         return newDeclaredReferenceExpression("this", pointerType, ctx.rawSignature, ctx)
     }
 
-    private val String.suffix: String
+    private val IASTLiteralExpression.valueWithSuffix: Pair<String, String>
         get() {
+            val value = String(this.value).lowercase(Locale.getDefault())
+
             var suffix = ""
 
-            // maximum suffix length is 3
+            // Maximum suffix length is 3
             for (i in 1..3) {
-                val digit = this.substring(max(0, this.length - i))
+                val digit = value.substring(max(0, value.length - i))
                 suffix =
                     if (
-                        digit.chars().allMatch { character ->
-                            character == 'u'.code ||
-                                character == 'l'.code ||
-                                character == 'f'.code ||
-                                character == 'd'.code
+                        digit.chars().allMatch {
+                            // We need to match on different suffixes, based on the kind of literal
+                            when (this.kind) {
+                                // See https://en.cppreference.com/w/cpp/language/floating_literal
+                                lk_float_constant -> it == 'f'.code || it == 'l'.code
+                                lk_integer_constant -> it == 'u'.code || it == 'l'.code
+                                else -> false
+                            }
                         }
                     ) {
                         digit
@@ -768,7 +771,8 @@ class ExpressionHandler(lang: CXXLanguageFrontend) :
                     }
             }
 
-            return suffix
+            // Supply the value with the suffix stripped, as well as the suffix
+            return Pair(value.substring(0, value.length - suffix.length), suffix)
         }
 }
 

--- a/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/cpp/ExpressionHandler.kt
+++ b/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/cpp/ExpressionHandler.kt
@@ -720,11 +720,20 @@ class ExpressionHandler(lang: CXXLanguageFrontend) :
     private fun handleFloatLiteral(ctx: IASTLiteralExpression): Expression {
         val (strippedValue, suffix) = ctx.valueWithSuffix
 
-        return when (suffix) {
-            "f" -> newLiteral(strippedValue.toFloat(), parseType("float"), ctx.rawSignature)
-            "l" ->
-                newLiteral(strippedValue.toBigDecimal(), parseType("long double"), ctx.rawSignature)
-            else -> newLiteral(strippedValue.toDouble(), parseType("double"), ctx.rawSignature)
+        return try {
+            when (suffix) {
+                "f" -> newLiteral(strippedValue.toFloat(), parseType("float"), ctx.rawSignature)
+                "l" ->
+                    newLiteral(
+                        strippedValue.toBigDecimal(),
+                        parseType("long double"),
+                        ctx.rawSignature
+                    )
+                else -> newLiteral(strippedValue.toDouble(), parseType("double"), ctx.rawSignature)
+            }
+        } catch (ex: NumberFormatException) {
+            // It could be that we cannot parse the literal, in this case we return an error
+            ProblemExpression("could not parse literal: ${ex.message}")
         }
     }
 

--- a/cpg-core/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/cpp/CXXLanguageFrontendTest.kt
+++ b/cpg-core/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/cpp/CXXLanguageFrontendTest.kt
@@ -843,9 +843,13 @@ internal class CXXLanguageFrontendTest : BaseTest() {
         assertIs<IntegerType>(hex.type)
         assertLocalName("unsigned long long", hex.type)
 
-        val duration = tu.variables["duration"]
-        assertNotNull(duration)
-        assertIs<ProblemExpression>(duration.initializer)
+        val duration_ms = tu.variables["duration_ms"]
+        assertNotNull(duration_ms)
+        assertIs<ProblemExpression>(duration_ms.initializer)
+
+        val duration_s = tu.variables["duration_s"]
+        assertNotNull(duration_s)
+        assertIs<ProblemExpression>(duration_s.initializer)
     }
 
     @Test

--- a/cpg-core/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/cpp/CXXLanguageFrontendTest.kt
+++ b/cpg-core/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/cpp/CXXLanguageFrontendTest.kt
@@ -781,55 +781,71 @@ internal class CXXLanguageFrontendTest : BaseTest() {
     @Test
     @Throws(Exception::class)
     fun testLiterals() {
-        val file = File("src/test/resources/literals.cpp")
-        val declaration = analyzeAndGetFirstTU(listOf(file), file.parentFile.toPath(), true)
-        val s = declaration.getDeclarationAs(0, VariableDeclaration::class.java)
-        assertEquals(createTypeFrom("char[]", true), s!!.type)
-        assertLocalName("s", s)
+        val file = File("src/test/resources/cxx/literals.cpp")
+        val tu = analyzeAndGetFirstTU(listOf(file), file.parentFile.toPath(), true)
+
+        val s = tu.variables["s"]
+        assertNotNull(s)
+        assertIs<PointerType>(s.type)
+        assertLocalName("char[]", s.type)
 
         var initializer = s.initializer as? Literal<*>
         assertNotNull(initializer)
         assertEquals("string", initializer.value)
 
-        val i = declaration.getDeclarationAs(1, VariableDeclaration::class.java)
-        assertEquals(createTypeFrom("int", true), i!!.type)
-        assertLocalName("i", i)
+        val i = tu.variables["i"]
+        assertNotNull(i)
+        assertIs<IntegerType>(i.type)
+        assertLocalName("int", i.type)
 
         initializer = i.initializer as? Literal<*>
         assertNotNull(initializer)
         assertEquals(1, initializer.value)
 
-        val f = declaration.getDeclarationAs(2, VariableDeclaration::class.java)
-        assertEquals(createTypeFrom("float", true), f!!.type)
-        assertLocalName("f", f)
+        val f = tu.variables["f"]
+        assertNotNull(f)
+        assertIs<FloatingPointType>(f.type)
+        assertLocalName("float", f.type)
 
         initializer = f.initializer as? Literal<*>
         assertNotNull(initializer)
         assertEquals(0.2f, initializer.value)
 
-        val d = declaration.getDeclarationAs(3, VariableDeclaration::class.java)
-        assertEquals(createTypeFrom("double", true), d!!.type)
-        assertLocalName("d", d)
+        val d = tu.variables["d"]
+        assertNotNull(d)
+        assertIs<FloatingPointType>(d.type)
+        assertLocalName("double", d.type)
 
         initializer = d.initializer as? Literal<*>
         assertNotNull(initializer)
         assertEquals(0.2, initializer.value)
 
-        val b = declaration.getDeclarationAs(4, VariableDeclaration::class.java)
-        assertEquals(createTypeFrom("bool", true), b!!.type)
-        assertLocalName("b", b)
+        val b = tu.variables["b"]
+        assertNotNull(b)
+        assertIs<BooleanType>(b.type)
+        assertLocalName("bool", b.type)
 
         initializer = b.initializer as? Literal<*>
         assertNotNull(initializer)
         assertEquals(false, initializer.value)
 
-        val c = declaration.getDeclarationAs(5, VariableDeclaration::class.java)
-        assertEquals(createTypeFrom("char", true), c!!.type)
-        assertLocalName("c", c)
+        val c = tu.variables["c"]
+        assertNotNull(c)
+        assertIs<IntegerType>(c.type)
+        assertLocalName("char", c.type)
 
         initializer = c.initializer as? Literal<*>
         assertNotNull(initializer)
         assertEquals('c', initializer.value)
+
+        val hex = tu.variables["hex"]
+        assertNotNull(hex)
+        assertIs<IntegerType>(hex.type)
+        assertLocalName("unsigned long long", hex.type)
+
+        val duration = tu.variables["duration"]
+        assertNotNull(duration)
+        assertIs<ProblemExpression>(duration.initializer)
     }
 
     @Test

--- a/cpg-core/src/test/resources/cxx/literals.cpp
+++ b/cpg-core/src/test/resources/cxx/literals.cpp
@@ -8,4 +8,5 @@ auto hex = 0xfull;
 
 // ms is a custom literal suffix (provided by <chrono>. however, we are not handling this at the moment, so this
 // will result in an error
-auto duration = 250ms;
+auto duration_ms = 250ms;
+auto duration_s = 2.5s;

--- a/cpg-core/src/test/resources/cxx/literals.cpp
+++ b/cpg-core/src/test/resources/cxx/literals.cpp
@@ -1,0 +1,11 @@
+char s[] = "string";
+int i = 1;
+float f = 0.2f;
+double d = 0.2;
+bool b = false;
+char c = 'c';
+auto hex = 0xfull;
+
+// ms is a custom literal suffix (provided by <chrono>. however, we are not handling this at the moment, so this
+// will result in an error
+auto duration = 250ms;

--- a/cpg-core/src/test/resources/literals.cpp
+++ b/cpg-core/src/test/resources/literals.cpp
@@ -1,6 +1,0 @@
-char s[] = "string";
-int i = 1;
-float f = 0.2f;
-double d = 0.2;
-bool b = false;
-char c = 'c';

